liberum_vault/src/vault.rs

```diff
@@ -0,0 +1,481 @@
+        const INSERT_FRAGMENT_QUERY: &str = "
+                INSERT INTO fragment (hash0, hash1, hash2, hash3, path, size)
+                VALUES (?1, ?2, ?3, ?4, ?5, ?6)
+            ";
+
+        let hash_as_u64 = fragment.hash.as_u64_slice_be();
+
+        let cnt = self
+            .db
+            .call(move |conn| {
+                let key_as_i64 = [
+                    hash_as_u64[0] as i64,
+                    hash_as_u64[1] as i64,
+                    hash_as_u64[2] as i64,
+                    hash_as_u64[3] as i64,
+                ];
+
+                let cnt = conn.query_row(SELECT_FRAGMENT_QUERY, key_as_i64, |r| {
+                    let cnt: usize = r.get(0)?;
+
+                    Ok(cnt)
+                })?;
+
+                Ok(cnt)
+            })
+            .await?;
+
+        if cnt != 0 {
+            return Ok(());
+        }
+
+        self.db
+            .call(move |conn| {
+                conn.execute(
+                    INSERT_FRAGMENT_QUERY,
+                    (
+                        hash_as_u64[0] as i64,
+                        hash_as_u64[1] as i64,
+                        hash_as_u64[2] as i64,
+                        hash_as_u64[3] as i64,
+                        fragment.path.to_str(),
+                        fragment.size,
+                    ),
+                )?;
+
+                Ok(())
+            })
+            .await
+            .map_err(|e| anyhow!(e))
+    }
+
+    async fn load_fragment(&self, key: Key) -> Result<Option<FragmentData>> {
+        let fragment_info = self.load_fragment_info(key.clone()).await?;
+
+        if let None = fragment_info {
+            return Ok(None);
+        }
+
+        let fragment_info = fragment_info.unwrap();
+        let fragment_path = fragment_info.path;
+        let fragment_file = File::open(&fragment_path).await?;
+
+        // TODO: Fix, is there any way to not use this take here?
+        Ok(Some(
+            ReaderStream::new(fragment_file.take(fragment_info.size)).boxed(),
+        ))
+    }
+
+    async fn store_fragment(&self, key: Option<Key>, data: &mut FragmentData) -> Result<Key> {
+        let uid = uuid::Uuid::new_v4();
+        let random_fragment_path = Self::temp_dir_path(&self.vault_dir_path).join(uid.to_string());
+        let mut fragment_file = File::create(&random_fragment_path).await?;
+        let mut hasher = blake3::Hasher::new();
+        let mut fragment_size = 0;
+
+        while let Some(bytes) = data.next().await {
+            let bytes = bytes?;
+            hasher.update(&bytes);
+            fragment_file.write(&bytes).await?;
+            fragment_size += bytes.len();
+        }
+
+        let key_bytes = hasher.finalize().as_bytes().to_vec();
+        let fragment_key = Key::try_from(key_bytes.clone())?;
+
+        // Verify integrity if key was provided
+        if let Some(key) = key {
+            if key != fragment_key {
+                bail!(
```

Comment:
We need to clean the file from the drive before returning. That's a (nonvolatile)memory leak :)

Reply:
Done